WordPress/src/main/java/org/wordpress/android/ui/posts/prepublishing/home/usecases/PublishPostImmediatelyUseCase.kt

```diff
@@ -0,0 +1,31 @@
+package org.wordpress.android.ui.posts.prepublishing.home.usecases
+
+import org.wordpress.android.fluxc.model.PostModel
+import org.wordpress.android.fluxc.model.post.PostStatus
+import org.wordpress.android.ui.posts.EditPostRepository
+import org.wordpress.android.ui.posts.EditPostRepository.UpdatePostResult
+import org.wordpress.android.util.DateTimeUtilsWrapper
+import javax.inject.Inject
+
+class PublishPostImmediatelyUseCase @Inject constructor(private val dateTimeUtilsWrapper: DateTimeUtilsWrapper) {
+    fun updatePostToPublishImmediately(
```

Comment:
Where was this logic located before? Ideally, we'd
1. re-use the original code and don't introduce a new use case 
2. remove the original code and use this class at the new location
3. refactor the original code so both the original and new location use this class

As we discussed, changing post statuses and date creation is very tricky and we need to be extremely careful not to introduce bugs.

Reply:
Thanks @malinajirka 

> re-use the original code and don't introduce a new use case
remove the original code and use this class at the new location
refactor the original code so both the original and new location use this class

This logic wasn't located anywhere. It's newly introduced due to the functionality needed.

> As we discussed, changing post statuses and date creation is very tricky and we need to be extremely careful not to introduce bugs.

Thanks for the reminder. I don't want to introduce any bugs at all 😅  

Let me know what you think about this solution. The class was introduced because when the user has a post that was scheduled and `Schedule` is the `PrimaryAction` when they click that action in the `Editor` you will see the bottom sheet. Within the bottom sheet, the `Publish Date` will have the date that's set so that the post can be published in the future. 

If you dismiss the bottom sheet and go back to the Editor, and press the more icon, the `Publish Now` Secondary Action is there. When the user presses this, I am expecting that they would want to see that the post is going to be published immediately because currently it's scheduled. The way to make `Immediately` is to clear the date and set the status. Let me know if this makes sense. I took this approach because we want the `EditPostRepository` to be the source of truth and once the user presses `Publish Now` I think it's okay that we clear the `Scheduled` `PostStatus`.

Reply:
Thanks for the explanation! 

>This logic wasn't located anywhere. It's newly introduced due to the functionality needed.

Someone had to change the status and clear the date even before we introduced the prepublishing nudges dialog, right? Wasn't it perhaps [this](https://github.com/wordpress-mobile/WordPress-Android/blob/a86967533294b8bdedb2160eefeb7aec2b92e227/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java#L1802) place? 

It seems the "EditPostActivity.uploadPost(.)` method is never invoked with `publish == true` now. Could we perhaps clean up the code there which would result in removing [this](https://github.com/wordpress-mobile/WordPress-Android/blob/a86967533294b8bdedb2160eefeb7aec2b92e227/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java#L1802) and I believe we'd follow the `2.` step from my previous comment ("remove the original code and use this class at the new location
"). Wdyt?


Can you please also elaborate why we want to change the status to "Draft" in certain scenarios? The use case is called `PublishPostImmediately`, so changing the status to "Draft" seems quite confusing. Wdyt? I'd at least suggest adding a comment explaining why we need it (if we really need it).

Reply:
Just to point it out here I discovered this method in the `PostUtils` 
https://github.com/wordpress-mobile/WordPress-Android/blob/b3d296e30c4d2da0ae0faefe0ebf348c0581650b/WordPress/src/main/java/org/wordpress/android/ui/posts/PostUtils.java#L549

I am thinking to reuse this for consistency.

Reply:
> Can you please also elaborate why we want to change the status to "Draft" in certain scenarios? The use case is called PublishPostImmediately, so changing the status to "Draft" seems quite confusing. Wdyt? I'd at least suggest adding a comment explaining why we need it (if we really need it).

Thanks for this question. I added a comment to make it clearer. When a post is created it's `PostStatus` is Draft. When it is `Draft` then Publish Now is shown. If it's Published then `Update Now` is shown. 

If the user has a post as `Scheduled` and it was previously `Draft` I thought it would be a bad experience to show them `Update Now` when `Publish Now` is pressed so this was the safest/clearest way I thought of doing it without additional logic elsewhere.

Reply:
> It seems the EditPostActivity.uploadPost(.) method is never invoked with publish == true now. Could we perhaps clean up the code there which would result in removing this and I believe we'd follow the 2. step from my previous comment ("remove the original code and use this class at the new location"). Wdyt?

Sounds good. It is passed with `true` here though. It just seems a bit hidden since it's getting the value from the `onClicked` event.

https://github.com/wordpress-mobile/WordPress-Android/blob/7148518e873fb95a5a167840a42168387feef1f7/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java#L1751

Reply:
> Could we perhaps clean up the code there which would result in removing this and I believe we'd follow the 2. step from my previous comment ("remove the original code and use this class at the new location"). Wdyt?

Thanks dcca109 85e732c

Reply:
>If the user has a post as Scheduled and it was previously Draft I thought it would be a bad experience to show them Update Now when Publish Now is pressed so this was the safest/clearest way I thought of doing it without additional logic elsewhere.

If I understand this correctly we are changing the status to "Draft" even though we want to publish the post just to show the correct title on the button? How would you feel about fixing the button label based on the `isNewPost` flag instead?

>Thanks dcca109 85e732c

Related note: PublishPostImmediatelyUseCase.updatePostToPublishImmediately has "publishPost" flag. Imagine you wouldn't want to look into the body of the class -> what do you think would happen if you called `updatePostToPublishImmediately` with `publishPost == false`? Tbh I'd expect the usecase wouldn't do anything at all in this scenario - it's called "updateToPublish" and I'm passing publish == false".

Reply:
> bh I'd expect the usecase wouldn't do anything at all in this scenario - it's called "updateToPublish" and I'm passing publish == false"

Exactly! This is spot on. The logic related to the button should can depend on the `isNewPost` 

> How would you feel about fixing the button label based on the isNewPost flag instead?

I agree with this.